### PR TITLE
Infer types from issubclass() calls

### LINKED ISSUE #2909: Types should be promoted by issubclass
I expected types to be promoted to sub-types within an `issubclass` block, the same as instances within an `isinstance` block, but type-checking the following code results in the error `Type[Vehicle] has no attribute "wheels"`:

```python
from typing import Type

class Vehicle(object):
    name = ''

class Car(Vehicle):
    wheels = 4

def test_isinstance(x):
    # type: (Vehicle) -> None
    print(x.name)
    if isinstance(x, Car):
        print(x.wheels)

def test_issubclass(x):
    # type: (Type[Vehicle]) -> None
    print(x.name)
    if issubclass(x, Car):
        print(x.wheels)
```

Is this be possible or am I missing something?  I'm testing against the master branch.

### PR DESCRIPTION
**Pull request by pkch**
Fixes #2909 

This PR includes a few commits from other PRs that haven't been merged in yet; I'll squash and fix any merge conflicts after the other PRs are merged in.

The only commits that are unique to this PR are the last 2: a4db99e and fe27fda, and the first of them is just a failing unit test.

The commit is somewhat messy, so I just wanted to get feedback on whether it's going in the right direction. Basically, I had to deal with the interplay of `Union` and `issubclass`. Everything seems to work, but it's not too elegant.

### CODE REVIEW

File: mypy/semanal.py [hunk 1]
```diff
@@ -2674,8 +2674,9 @@ def visit_member_expr(self, expr: MemberExpr) -> None:
                 # one type checker run. If we reported errors here,
                 # the build would terminate after semantic analysis
                 # and we wouldn't be able to report any type errors.
-                full_name = '%s.%s' % (file.fullname() if file is not None else None, expr.name)
-                mod_name = " '%s'" % file.fullname() if file is not None else ''
+                full_name = '%s.%s' % (file.fullname() if file is not None  # type: ignore
+                                       else None, expr.name)
+                mod_name = " '%s'" % file.fullname() if file is not None else ''  # type: ignore
```
Comment (ilevkivskyi) [inline mypy/semanal.py]: Why did you add three ``# type: ignore``s in this file? Are you trying to self check mypy with ``--strct-optional``?
If yes, then i think you should not put them here in this PR.

File: test-data/unit/check-isinstance.test [hunk 2]
```diff
@@ -1344,3 +1351,131 @@ def f(x: object) -> None:
...
 [out]
+
+[case testIsInstanceWithTypeVariable]
+from typing import *
+def f(x: Union[int, str], typ: type) -> None:
+    if isinstance(x, (typ, int)):
+        x + 1  # E: Unsupported operand types for + (likely involving Union)
+[builtins fixtures/isinstancelist.pyi]
+
+[case testIssubclass]
```
Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: Is it possible to split this test into few smaller unit tests?

File: test-data/unit/check-isinstance.test [hunk 3]
```diff
@@ -31,6 +31,13 @@ z = [a.y for b in y for a in b]
 
 [builtins fixtures/list.pyi]
 
+[case testIsinstanceNestedTuple]
+from typing import Union, List, Tuple
+def f(x: Union[int, str, List, Tuple]) -> None:
+    if isinstance(x, (str, (int, tuple))):
+        x[1]  # E: Value of type "Union[int, str, tuple]" is not indexable
```
Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: I would also add a similar test with a successful check (i.e. not causing an error).

File: test-data/unit/check-isinstance.test [hunk 4]
```diff
@@ -1344,3 +1351,131 @@ def f(x: object) -> None:
     reveal_type(b)  # E: Revealed type is '__main__.A'
 [builtins fixtures/isinstance.pyi]
 [out]
+
+[case testIsInstanceWithTypeVariable]
```
Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: The name of this test is misleading. You don't use type variables here.

File: test-data/unit/check-isinstance.test [hunk 5]
```diff
@@ -1344,3 +1351,131 @@ def f(x: object) -> None:
...
+    else:
+        cls.aggressive
+        cls.level
+        cls.job  # E: Type[Goblin] has no attribute "job"
+        g = cls()
+        g.level = 15
+        g.job  # E: "Goblin" has no attribute "job"
+
+
+[builtins fixtures/isinstancelist.pyi]
```
Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: I would add also some tests with built-in types, e.g. something involving
```python
class MyList(List[int]): ...
```
and then
```python
if issuclass(MyList, List): ...
```

File: mypy/checker.py [hunk 6]
```diff
@@ -2693,12 +2723,31 @@ def flatten(t: Expression) -> List[Expression]:
...
-        all_types = type.items
+    assert len(types) != 0
+    if len(types) == 1:
+        return types[0]
     else:
-        all_types = [type]
+        return UnionType(types)
+
+
+def get_issubclass_type(expr: Expression, type_map: Dict[Expression, Type]) -> Type:
```
Comment (pkch) [inline mypy/checker.py]: Ah I thought it might be similar, but then I actually managed to make it work without changing anything at all - and didn't notice. I removed it.

File: mypy/checker.py [hunk 7]
```diff
@@ -2693,12 +2723,31 @@ def flatten(t: Expression) -> List[Expression]:
...
+    all_types = [type_map[e] for e in flatten(expr)]
+
+    types = []  # type: List[Type]
+
+    for type in all_types:
+        if isinstance(type, FunctionLike):
+            if type.is_type_obj():
+                # Type variables may be present -- erase them, which is the best
+                # we can do (outside disallowing them here).
+                type = erase_typevars(type.items()[0].ret_type)
```
Comment (pkch) [inline mypy/checker.py]: I should probably do it in a separate PR though? (Since I deleted my function, your comment now applies only to `get_isinstance_type`, which I'm not modifying.) Or is it ok to incorporate it here as well?

File: mypy/semanal.py (hunk 1 above)
Comment (pkch) [inline mypy/semanal.py]: No, it's because of #2999. A bug fix in this commit triggered that issue. @JukkaL suggested to replace `#  type: ignore` with `cast` to make it safer, so I'll do that.

**Comment by pkch** [comment]
@ilevkivskyi  I addressed all your comments, except https://github.com/python/mypy/pull/3005#discussion_r106779110 (I replied). 

I've never had so many merge mistakes (not just conflicts, but actual mistakes...) 😞  I'll never again create a PR that depends on 2 unmerged PRs...

File: mypy/checker.py (hunk 7 above)
Comment (ilevkivskyi) [inline mypy/checker.py]: I would prefer a separate PR for this.

**Comment by ilevkivskyi** [comment]
@pkch 
> I had a huge number of merge mistakes (not just conflicts, but actual mistakes...) :disappointed: I'll never again create a PR that depends on 2 unmerged PRs...

This is not a big deal. But from organizational point of view, yes, independent PRs are better.

It looks like both PRs you mentioned are merged now. I am waiting to review your new commits (also a merge conflict needs to be fixed).

File: mypy/checker.py [hunk 8]
```diff
@@ -2588,6 +2588,20 @@ def or_conditional_maps(m1: TypeMap, m2: TypeMap) -> TypeMap:
     return result
 
 
+def convert_to_types(m: MutableMapping[Expression, Type]) -> None:
+    for k in m:
```
Comment (pkch) [inline mypy/checker.py]: done

File: mypy/checker.py [hunk 9]
```diff
@@ -2613,6 +2627,22 @@ def find_isinstance_check(node: Expression,
                 vartype = type_map[expr]
                 type = get_isinstance_type(node.args[1], type_map)
                 return conditional_type_map(expr, vartype, type)
+        elif refers_to_fullname(node.callee, 'builtins.issubclass'):
+            expr = node.args[0]
+            if expr.literal == LITERAL_TYPE:
+                vartype = type_map[expr]
+                type = get_issubclass_type(node.args[1], type_map)
+                if isinstance(vartype, UnionType):
+                    vartype = UnionType([t.item for t in vartype.items])  # type: ignore
```
Comment (pkch) [inline mypy/checker.py]: Fixed using isinstance

File: mypy/checker.py (hunk 7 above)
Comment (pkch) [inline mypy/checker.py]: I added #3037 ; not sure where to actually catch this problem.

File: test-data/unit/check-isinstance.test [hunk 10]
```diff
@@ -1412,3 +1412,171 @@ def f(x: Union[int, A], a: Type[A]) -> None:
...
+from typing import Type, ClassVar
+
+class Goblin:
+    aggressive: bool = True
+    level: int
+
+class GoblinAmbusher(Goblin):
+    job: ClassVar[str] = 'Ranger'
+
+
```
Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: One blanc line is enough here

File: mypy/semanal.py (hunk 1 above)
Comment (ilevkivskyi) [inline mypy/semanal.py]: This is still not done.

File: mypy/checker.py [hunk 11]
```diff
@@ -2639,8 +2653,30 @@ def find_isinstance_check(node: Expression,
...
+                vartype = type_map[expr]
+                type = get_isinstance_type(node.args[1], type_map)
+                if isinstance(vartype, UnionType):
+                    union_list = []
+                    for t in vartype.items:
+                        if isinstance(t, TypeType):
+                            union_list.append(t.item)
+                        else:
+                            #  this an error; should be caught earlier, so we should never be here
+                            return {}, {}
```
Comment (ilevkivskyi) [inline mypy/checker.py]: If you are sure this is an unreachable branch, then you could put ``assert False`` here. If this could happen as a result of an error that is reported elsewhere, then adjust the comment correspondingly.

File: mypy/checker.py [hunk 12]
```diff
@@ -2639,8 +2653,30 @@ def find_isinstance_check(node: Expression,
...
+            if expr.literal == LITERAL_TYPE:
+                vartype = type_map[expr]
+                type = get_isinstance_type(node.args[1], type_map)
+                if isinstance(vartype, UnionType):
+                    union_list = []
+                    for t in vartype.items:
+                        if isinstance(t, TypeType):
+                            union_list.append(t.item)
+                        else:
+                            #  this an error; should be caught earlier, so we should never be here
```
Comment (ilevkivskyi) [inline mypy/checker.py]: typo: "this an error;" -> "this is an error;"

File: test-data/unit/check-isinstance.test [hunk 13]
```diff
@@ -1412,3 +1412,171 @@ def f(x: Union[int, A], a: Type[A]) -> None:
...
+from typing import List, Type
+
+class MyList(List): pass
+class MyIntList(List[int]): pass
+
+def f(cls: Type):
+    if issubclass(cls, MyList):
+        cls()[0]
+    else:
+        cls()[0]  # E:
```
Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: What is the error message here? Why didn't you write it here?

File: test-data/unit/check-isinstance.test [hunk 14]
```diff
@@ -1412,3 +1412,171 @@ def f(x: Union[int, A], a: Type[A]) -> None:
 
 [builtins fixtures/isinstancelist.pyi]
 
+
+[case testIssubclass]
+from typing import Type, ClassVar
+
+class Goblin:
+    aggressive: bool = True
+    level: int
```
Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: Do you need both attributes here? Some test cases could be simplified if you have only one instance variable.
Also you probably don't need to check both access and assignment to instance variables in all tests, just checking assignment is enough.

File: test-data/unit/check-isinstance.test (hunk 13 above)
Comment (pkch) [inline test-data/unit/check-isinstance.test]: Oh, no error message I just accidentally left `E:`

Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: But shouldn't this actually be an error? In the ``else`` branch, ``cls`` is not ``List`` and therefore not subscriptable. Or am I missing something?

Comment (pkch) [inline test-data/unit/check-isinstance.test]: No, it's my bad. I confused this test with another where I left `E:`; in this case I did mean for it to return an error, and it doesn't, and I forgot about it.

The reason it doesn't complain is that `cls()` is revealed as `Any`, since `cls` is defined as `Type`, which is translated to `Type[Any]`. So the entire test was useless. 

Btw, I would prefer `Type` to be expanded to `Type[object]` rather than `Type[Any]`, but I suppose it would be too inconsistent.

File: mypy/semanal.py [hunk 15]
```diff
@@ -2696,7 +2696,8 @@ def visit_member_expr(self, expr: MemberExpr) -> None:
                 # one type checker run. If we reported errors here,
                 # the build would terminate after semantic analysis
                 # and we wouldn't be able to report any type errors.
-                full_name = '%s.%s' % (file.fullname() if file is not None else None, expr.name)
+                full_name = '%s.%s' % (file.fullname() if file is not None
+                                       else None, expr.name)
```
Comment (ilevkivskyi) [inline mypy/semanal.py]: I think you don't need this formatting change, it only creates noise in the PR. Otherwise now it looks good to me

**Comment by ilevkivskyi** [comment]
@pkch I know that I already approved this PR but there are two more things that could be fixed before I merge this:
* most tests should use ``reveal_type`` in ``if issubclas ... else: ...`` branches (this will probably also allow further simplifying some tests).
* there are two ``TODO`` items that should be fixed.

**Comment by ilevkivskyi** [comment]
@pkch Sorry, I cannot merge this yet. This PR _crashes_ with:
```python
x: Type[str]
if issubclass(x, int):
    reveal_type(x)
```
The reason is that ``conditional_type_map`` sometimes returns ``None`` (for example for non-overlapping case like in the example above) and this case should be carefully treated.

**Comment by pkch** [comment]
Ah yes. BTW, this is the type of bug that `mypy --strict-optional mypy` would have caught.

**Comment by gvanrossum** [comment]
Speaking of which, we should really convert all of mypy to strict optional checking. (It's not so easy, or we would have done it already.)

File: test-data/unit/check-isinstance.test [hunk 16]
```diff
@@ -1423,31 +1423,40 @@ def f(x: Union[int, A], a: Type[A]) -> None:
...
-        ga.level = 15
-        ga.job
-        ga.job = "Warrior"  # E: Cannot assign to class variable "job" via instance
     else:
         reveal_type(cls)  # E: Revealed type is 'Type[__main__.Goblin]'
-        cls.level
-        cls.job  # E: Type[Goblin] has no attribute "job"
-        g = cls()
-        g.level = 15
-        g.job  # E: "Goblin" has no attribute "job"
```
Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: I think few integration tests would be useful, so I was happy with these tests. Why are you removing them in this commit?

**Comment by ilevkivskyi** [comment]
@gvanrossum could you please try this PR with your internal codebase, just to be sure that everything is OK?

I think it is a useful feature and the PR is almost ready to be merged (after @pkch will restore the tests that were inadvertently removed).

File: test-data/unit/check-isinstance.test (hunk 16 above)
Comment (pkch) [inline test-data/unit/check-isinstance.test]: Ah I thought the `reveal_type` makes for a much more focused unit test. I was afraid the test was needlessly long and hard to read. After all, this PR is only for `issubclass` functionality, once we know that the revealed type is correct, do we really need to check that the attribute access is done correctly? But I'm adding back a couple integration tests.

Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: Sometimes mypy does ``copy_modified``, so same representation of type is not a 100% guarantee. Also, there were some tests with destructuring unions in the initial version, I could not find them now, something like:
```
x: Type[Union[int, str, X]]
if isinstance(x, (int, str)):
    reveal_type(x)
    if isinstance(x, str):
        reval_type(x)
    else:
        reval_type(x)
else:
    reval_type(x)  # Revealed type is '__main__.X'
```
etc.

Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: > same representation of type is not a 100% 

TBH, I am only a bit worried about ``ClassVar``, but still few other integration tests will be useful.

**Comment by gvanrossum** [comment]
This is clean on our internal codebase (FWIW).

File: test-data/unit/check-isinstance.test (hunk 16 above)
Comment (pkch) [inline test-data/unit/check-isinstance.test]: Hmm are you sure by "destructuring unions" you didn't mean this:

```
[case testIsinstanceNestedTuple]
from typing import Union, List, Tuple, Dict
def f(x: Union[int, str, List]) -> None:
    if isinstance(x, (str, (int,))):
        reveal_type(x)  # E: Revealed type is 'Union[builtins.int, builtins.str]'
        x[1]  # E: Value of type "Union[int, str]" is not indexable
    else:
        reveal_type(x)  # E: Revealed type is 'builtins.list[Any]'
        x[1]
    reveal_type(x)  # E: Revealed type is 'Union[builtins.int, builtins.str, builtins.list[Any]]'
    if isinstance(x, (str, (list,))):
        reveal_type(x)  # E: Revealed type is 'Union[builtins.str, builtins.list[Any]]'
        x[1]
    reveal_type(x)  # E: Revealed type is 'Union[builtins.int, builtins.str, builtins.list[Any]]'
[builtins fixtures/isinstancelist.pyi]
```

I don't recall adding it to `issubclass`, only to `isinstance` PR.

Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: Ah yes, this was for ``isinstance()``, could you add a similar test for ``issubclass()`` (as I have proposed above)? Sorry, you are probably already tired, but I would like to be safe here.

Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: > Ah yes, this was for isinstance(), could you add a similar test for ``issubclass()`` (as I have proposed above)? Sorry, you are probably already tired, but I would like to be safe here.

@pkch Just in case you didn't get notification.

Comment (pkch) [inline test-data/unit/check-isinstance.test]: Heh, np, this is a useful test. A bunch of weird stuff but I think it's tangential to the main purpose of this PR:

- `Type[List]` is revealed differently in different locations: sometimes `"Type[builtins.list]"`, sometimes `"Type[builtins.list[Any]]"`
- Instantiation of that is revealed as `builtins.list[<uninhabited>]`

Other than revealed type strings, everything looks good. Should I open an issue for those two, or should we pretend we didn't notice?

(the second one is probably going to be fixed with #3032)

File: test-data/unit/check-isinstance.test [hunk 17]
```diff
@@ -1443,6 +1443,28 @@ else:
 [builtins fixtures/isinstancelist.pyi]
 
 
+[case testIssubclasDestructuringUnions]
+from typing import Union, List, Tuple, Dict, Type
+def f(x: Union[Type[int], Type[str], Type[List]]) -> None:
```
Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: @pkch Is it possible to use ``Type[Union[int, str, List]]`` here?

File: test-data/unit/check-isinstance.test (hunk 16 above)
Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: @pkch 
I think this is a separate issue related to what we discussed in tuple fallback issue #3117. I think we should weed out all generic types with incorrect type argument count, there should be no such thing as ``builtins.list``, only ``builtins.list[AnyType()]``, etc. internally. Otherwise such inconsistencies lead to weird bugs like strange ``<uninhabited>`` or the problem with tuple

File: test-data/unit/check-isinstance.test (hunk 17 above)
Comment (pkch) [inline test-data/unit/check-isinstance.test]: Wow, it is, but I thought it's illegal? I thought `Type` requires only simple types inside, not constructs like `Union` etc.?

But it *50%* works... Here's the output. The first couple lines `int` and `str` get swapped; then a couple lines nothing changed; and then it completely breaks down and becomes `Any`:

```
def f(x: Type[Union[int, str, List]]) -> None:
    if issubclass(x, (str, (int,))):
        reveal_type(x)  # E: Revealed type is 'Union[Type[builtins.str], Type[builtins.int]]'
        reveal_type(x())  # E: Revealed type is 'Union[builtins.str, builtins.int]'
        x()[1]  # E: Value of type "Union[str, int]" is not indexable
    else:
        reveal_type(x)  # E: Revealed type is 'Type[builtins.list]'
        reveal_type(x())  # E: Revealed type is 'builtins.list[<uninhabited>]'
        x()[1]
    reveal_type(x)  # E: Revealed type is 'Any'
    reveal_type(x())  # E: Revealed type is 'Any'
    if issubclass(x, (str, (list,))):
        reveal_type(x)  # E: Revealed type is 'Any'
        reveal_type(x())  # E: Revealed type is 'Any'
        x()[1]
    reveal_type(x)  # E: Revealed type is 'Any'
    reveal_type(x())  # E: Revealed type is 'Any'
```

Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: @pkch 
>  I thought ``Type`` requires only simple types inside, not constructs like ``Union``

``Union`` is legal as a special case, see for example [typing docs](https://docs.python.org/3/library/typing.html#typing.Type) (emphasis my):
> The only legal parameters for ``Type`` are classes, _unions of classes_, and ``Any``.

> But it 50% works...

Hmm, these two ``Type[Union[int, str]]`` and ``Union[Type[int], Type[str]]`` should be equivalent I think. Is it easy to fix?

File: test-data/unit/check-isinstance.test (hunk 16 above)
Comment (pkch) [inline test-data/unit/check-isinstance.test]: Sadly, my #3129 doesn't achieve that unification of tuple fallbacks in this case (I was hoping it might, but when I merged that PR into here, nothing changed). Should this be done in this PR or in a separate one?

Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: This could be done in #3129 or in an additional PR, but _not here_.

File: test-data/unit/check-isinstance.test (hunk 17 above)
Comment (pkch) [inline test-data/unit/check-isinstance.test]: Wow `reveal_type` changes from early to late in the same scope:

```
from typing import Union, List, Tuple, Dict, Type
def f(x: Type[Union[int, str, List]]) -> None:
    reveal_type(x)  # E: Revealed type is 'Type[Union[builtins.int, builtins.str, builtins.list]]'
    reveal_type(x())  # E: Revealed type is 'Union[builtins.int, builtins.str, builtins.list[<uninhabited>]]'
    if issubclass(x, (str, (int,))):
        1
    else:
        1
    reveal_type(x)  # E: Revealed type is 'Any'
    reveal_type(x())  # E: Revealed type is 'Any'
```

Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: I think this should be fixed. Without this PR I get:
```python
def f(x: Type[Union[int, str, List]]) -> None:
    reveal_type(x)  # E: Revealed type is 'Type[Union[builtins.int, builtins.str, builtins.list]]'
    if issubclass(x, (str, (int,))):
        1
    else:
        1
    reveal_type(x)  # E: Revealed type is 'Type[Union[builtins.int, builtins.str, builtins.list]]'
```
A type should not be "destroyed" after an ``issubcheck()`` check.

Comment (pkch) [inline test-data/unit/check-isinstance.test]: I fixed it by choosing one canonical internal representation for both `Union[Type[...]]` and `Type[Union[...]]` . Instances of `TypeType` now have to created using `.make()` static method, which will check if it should return `UnionType` or `TypeType`. It's very reliable because I disallowed the use of `TypeType` constructor (otherwise, we can't prevent someone in the future accidentally creating the wrong internal representation).

As a result, this commit accidentally fixed a broken test, `testTypeUsingTypeCClassMethodUnion`.

Unfortunately, it also means a slight annoyance of having to use a class method to create `TypeType`.

The alternative is for me to renaming class `TypeType` into `_TypeType`, and exposing a function `TypeType`. This will completely keep the current API. But I'm afraid a lot of type hints would break because functions and classes aren't equivalent for purposes of typing, even if they have the same signature). But lmk if you prefer this approach.

Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: TBH, this looks to big a hammer for this problem. I would propose to make your last commit a separate PR and a corresponding issue for discussion. Then I will merge this PR

Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: @pkch 
> Then I will merge this PR

This time for sure :-)

Comment (pkch) [inline test-data/unit/check-isinstance.test]: Done! 

Interesting, when I `git reset --hard @~` and `git push -f`, I effectively remove the last commit from this PR. Then AppVeyor decided to rebuild, while Travis decided not to. At first I thought that AppVeyor was smarter because it noticed that master branch changed so it figured the new merge with master will be different, and it's worth rebuilding.

It turns out that AppVeyor rebuilds regardless of whether master changed since the last time it built the commit; in fact, it even rebuilt on my own branch in the forked repo.

So, when they already built for the commit that just became the head of the PR or branch:

- AppVeyor always rebuilds, even if there's absolutely no need
- Travis does not rebuild, even if it's arguably worth rebuilding due to the change in master

Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: @pkch 

Something strange happened with typeshed. Could you please fix this?
(And don't forget to open an issue about ``Type[Union[...]]`` vs ``Union[Type[...]]``.)

Comment (pkch) [inline test-data/unit/check-isinstance.test]: @ilevkivskyi what do you see happening? everything seems fine in my forked repo, when I merged this branch into master.

Comment (ilevkivskyi) [inline test-data/unit/check-isinstance.test]: @pkch 
Hm, this is strange, GitHub shows me that your PR makes changes to typeshed, but I checked the commit hash, and it is the correct one. I think it is safe to merge.
